packages/rendermime/src/renderers.ts

```diff
@@ -801,16 +819,57 @@ export function renderText(options: renderText.IRenderOptions): Promise<void> {
 
   const preTextContent = pre.textContent;
 
+  let cacheStoreOptions = [];
+  if (autoLinkOptions.checkWeb) {
+    cacheStoreOptions.push('web');
+  }
+  if (autoLinkOptions.checkPaths) {
+    cacheStoreOptions.push('paths');
+  }
+  const cacheStoreKey = cacheStoreOptions.join('-');
+  let cacheStore = Private.autoLinkCache.get(cacheStoreKey);
+  if (!cacheStore) {
+    cacheStore = new WeakMap();
+    Private.autoLinkCache.set(cacheStoreKey, cacheStore);
+  }
+
   let ret: HTMLPreElement;
   if (preTextContent) {
     // Note: only text nodes and span elements should be present after sanitization in the `<pre>` element.
-    const linkedNodes =
-      sanitizer.getAutolink?.() ?? true
-        ? autolink(preTextContent, {
-            checkWeb: true,
-            checkPaths: false
-          })
-        : [document.createTextNode(content)];
+    let linkedNodes: (HTMLAnchorElement | Text)[];
+    if (sanitizer.getAutolink?.() ?? true) {
+      const cache = getApplicableLinkCache(
+        cacheStore.get(host),
+        preTextContent
+      );
+      if (cache) {
+        const { cachedNodes, addedText } = cache;
+        const fromCache = cachedNodes;
+        const newAdditions = autolink(addedText, autoLinkOptions);
+        const lastInCache = fromCache[fromCache.length - 1];
+        const firstNewNode = newAdditions[0];
+
+        if (lastInCache instanceof Text && firstNewNode instanceof Text) {
+          const joiningNode = lastInCache;
+          joiningNode.data += firstNewNode.data;
+          linkedNodes = [
+            ...fromCache.slice(0, -1),
+            joiningNode,
+            ...newAdditions.slice(1)
+          ];
+        } else {
+          linkedNodes = [...fromCache, ...newAdditions];
+        }
+      } else {
+        linkedNodes = autolink(preTextContent, autoLinkOptions);
+      }
+      cacheStore.set(host, {
+        preTextContent,
+        linkedNodes
+      });
```

Comment:
Here is the problem: when search highlights are applied, they modify the nodes in the DOM by calling [`Text.splitText`](https://developer.mozilla.org/en-US/docs/Web/API/Text/splitText):

https://github.com/jupyterlab/jupyterlab/blob/d317b2e1b728578bbb6959a412e7be485c987110/packages/documentsearch/src/providers/genericsearchprovider.ts#L317

since the same nodes are in cache, the spliced version ends up inserted next time. If search string covers the entire text then it seems as if nothing was added for subsequent renders where cache is hit.

If we add `.cloneNode()` for each of `linkedNodes` then the performance drops to the same level as before this patch.

Reply:
We can fix this by cloning the node in search highlights code instead, but this would impact the performance of highlighting matches.

Ideally, we would only clone cache entries on Node's modification (well, prior to it). This is not easily possible.

Reply:
Maybe instead of storing the DOM elements in the cache we could serialize the recipes for those; this would be easy because we have only text and anchors to deal with...

Reply:
Looking at it again it seems that there is still a significant performance gain even when calling `.cloneNode()` each time. I will update the benchmaks later.

Reply:
Verified both memory usage and performance, neither is significantly affected

For new lines scenario (standard output) where nodes can be collapsed into a single node the overhead of calling `.cloneNode(true)` is negligible:

![image](https://github.com/user-attachments/assets/8207dbfc-b120-4735-821d-02220fdc8228)

For the scenario with many nodes (standard output example) the overhead adds up to ~1s over when running:

```python
import sys

for i in range(1000):
    print(f"example.com {i} XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX www.", end='')
    sys.stdout.flush()
```

which took 7 seconds with `.deepClone(true)` as compared to 14 seconds before this PR.

![image](https://github.com/user-attachments/assets/dcaab07e-3e61-4301-ae09-f5edc2eeba86)

Note the timings are different than in the top-level comment as I am testing on a different hardware configuration today.